markets/perps-market/test/integration/Orders/OffchainAsyncOrder.fees.test.ts

```diff
@@ -1,6 +1,420 @@
-// TODO
-// Use this test to check correct fees are paid for offchain orders
-// to keeper
-// to market
-// to spot (exchange fee)
-// to perps (funding)
+import { ethers } from 'ethers';
+import { DEFAULT_SETTLEMENT_STRATEGY, bn, bootstrapMarkets, decimalMul } from '../bootstrap';
+import { fastForwardTo } from '@synthetixio/core-utils/utils/hardhat/rpc';
+import { snapshotCheckpoint } from '@synthetixio/core-utils/utils/mocha/snapshot';
+import { SynthMarkets } from '@synthetixio/spot-market/test/common';
+import {
+  DepositCollateralData,
+  OpenPositionData,
+  depositCollateral,
+  openPosition,
+  settleOrder,
+} from '../helpers';
+import assertBn from '@synthetixio/core-utils/utils/assertions/assert-bignumber';
+import { getTxTime } from '@synthetixio/core-utils/src/utils/hardhat/rpc';
+
+describe('Offchain Async Order test - fees', () => {
+  const orderFees = {
+    makerFee: bn(0.0003), // 3bps
+    takerFee: bn(0.0008), // 8bps
+  };
+  const ethPrice = bn(1000);
+
+  const { systems, perpsMarkets, synthMarkets, provider, trader1, keeper } = bootstrapMarkets({
+    synthMarkets: [
+      {
+        name: 'Bitcoin',
+        token: 'snxBTC',
+        buyPrice: bn(10_000),
+        sellPrice: bn(10_000),
+      },
+    ],
+    perpsMarkets: [
+      {
+        name: 'Ether',
+        token: 'snxETH',
+        price: ethPrice,
+        // setting to 0 to avoid funding and p/d price change affecting pnl
+        fundingParams: { skewScale: bn(0), maxFundingVelocity: bn(0) },
+        orderFees,
+      },
+    ],
+    traderAccountIds: [2, 3],
+  });
+  let ethMarketId: ethers.BigNumber;
+  let btcSynth: SynthMarkets[number];
+
+  before('identify actors', async () => {
+    ethMarketId = perpsMarkets()[0].marketId();
+    btcSynth = synthMarkets()[0];
+  });
+
+  const restoreToDepositCollateral = snapshotCheckpoint(provider);
+
+  const testCases: Array<{ name: string; collateralData: DepositCollateralData }> = [
```

Comment:
the type here will generally be implied based on the data so no need to explicitly set it, but either way non issue, mostly fyi :)